### ISSUE TITLE
Refresh token

### ISSUE THREAD

**Issue by JohnSmall** [minimul/qbo_api#59]
Is there any method in this gem that allows me to send a refresh token to QBO and get back a fresh access token?

From what I can see I have to handle 401 errors and create my own method to get a new access token. Since that's a pretty common thing to have to do, it's would make sense if it was part of the gem. 

Am I missing something?

**Comment by minimul** [comment]
I haven't had a OAuth2 gig yet and that is only scenario where this would be relevant.

With OAuth1 and its long validation duration you just use a cron job to auto-renew tokens.
Moreover, in OAuth1 if get an unauthenticated error you've missed your renew window.

You can do the [same](https://minimul.com/refreshing-the-quickbooks-oauth2-access-token.html) in OAuth2 but the method that you are suggesting is probably superior and is what Intuit is doing in their SDKs for OAuth2 (catching auth failure and then renewing on the fly).

I am open to solutions. Care to take a crack at it?

**Comment by JohnSmall** [comment]
The key issue is storing the access token. So I'm thinking of doing something that yields to a block so people can easily plug in their own storage solution. I'll do PR when I've done it.

**Comment by minimul** [comment]
Right. Good idea.

**Comment by bf4** [comment]
I've implemented it. I could make a PR

**Comment by minimul** [comment]
@bf4 Yes, if you could.

**Comment by bf4** [comment]
@minimul I just released a gem for handling commands that may fail.  Would you be ok with me adding it, or should I just implement the pattern in qbo_api? https://github.com/bf4/safety_dance

Might not make the PR today right now my code is something like

```ruby
      # Fails badly if entity is nil
      def get(entity, id, params: nil)
        params = configure_params(params)
        handle_errors { quickbooks_client.get(entity, id, params: params) }
      end

      def handle_errors(&block)
        result = ::Result.new(&block)
        return result if result.ok?
        case result.error
        when QboApi::Unauthorized
          QboApi.logger.info "Quickbooks access_token has expired. Refreshing and trying again"
          @quickbooks_client = refresh_client(quickbooks_client)
          ::Result.new(&block).tap do |new_result|
            if new_result.ok?
              QboApi.logger.info "Quickbooks access_token has been refreshed and request retry succeeded: #{new_result}"
            else
              notify_retry_failed(result, new_result)
            end
          end
        else
          result
        end
      end

    def refresh_client(quickbooks_client)
      refresh_access_token
      QboApi.new(access_token: current_access_token, realm_id: realm_id)
    end

    def refresh_access_token(current_refresh_token: refresh_token)
      fail ArgumentError, "refresh_token cannot be nil" if current_refresh_token.nil?
      url = 'https://oauth.platform.intuit.com'
      token_endpoint = '/oauth2/v1/tokens/bearer'

      connection = Faraday.new(url: url) { |conn|
        conn.basic_auth(client_id, client_secret)
        conn.request :url_encoded # application/x-www-form-urlencoded
        conn.headers['Accept'] = "application/json"

        conn.use FaradayMiddleware::ParseJson, parser_options: { symbolize_names: true }
        conn.use Faraday::Response::RaiseError
        conn.response :detailed_logger, QboApi.logger if QboApi.log
        conn.adapter  Faraday.default_adapter
      }
      raw_response = connection.post {|req|
        req.body = { grant_type: :refresh_token, refresh_token: current_refresh_token }
        req.url token_endpoint
      }

      data = raw_response.body
      # {
      #   "expires_in": 3600,
      #   "x_refresh_token_expires_in": 8706412,
      #   "refresh_token": "newnew...",
      #   "access_token": "new...",
      #   "token_type": "bearer"
      # }
      new_refresh_token = data.fetch(:refresh_token)
      if new_refresh_token != current_refresh_token
        QboApi.logger.info { "Quickbooks RefreshToken changed to '#{new_refresh_token}' from '#{current_refresh_token}'" }
      end
      self.refresh_token = new_refresh_token
      self.access_token = data.fetch(:access_token)
      data
    end

    def refresh_token
      cache.fetch("quickbooks.refresh_token")
    end

    def realm_id
      ENV.fetch("QUICKBOOKS_REALM_ID")
    end

    def client_id
      ENV.fetch("QUICKBOOKS_CLIENT_ID")
    end

    def secret_id
      ENV.fetch("QUICKBOOKS_SECRET_ID")
    end

    def cache
      @cache ||= defined?(Rails) ? Rails.cache : ActiveSupport::Cache::MemoryStore.new
    end


```

**Comment by minimul** [comment]
I am starting to think that this approach to renewing OAuth2 tokens should be its own gem. It would require perhaps a small modification to `qbo_api` to enable this new gem to hook into `QboApi::Unauthorized`

Thoughts?

/cc @ntippie

**Comment by ntippie** [comment]
I haven't implemented OAuth2 for QBO either, but perhaps using Faraday middleware would be a good solution? I found [this](https://github.com/grokify/faraday_middleware-oauth2_refresh/blob/master/lib/faraday_middleware/oauth2_refresh.rb), which looks similar to what I was imagining. However, that only refreshes if the token has already expired and doesn't catch HTTP 401.

**Comment by bf4** [comment]
For posterity and reference, pasting the code @ntippie referenced

```ruby
require 'base64'
require 'faraday'
require 'forwardable'

module FaradayMiddleware
  class OAuth2Refresh < Faraday::Middleware
    AUTH_HEADER = 'Authorization'.freeze

    attr_reader :oauth2_token

    extend Forwardable

    def call(env)
      if @oauth2_token.expired?
        @oauth2_token = @oauth2_token.refresh!({ headers: { 'Authorization' => 'Basic ' + get_api_key() } })
      end
      
      unless @oauth2_token.token.to_s.empty?
        env[:request_headers][AUTH_HEADER] = %(Bearer #{@oauth2_token.token})
      end

      @app.call env
    end

    def get_api_key
      api_key = Base64.encode64("#{@oauth2_token.client.id}:#{@oauth2_token.client.secret}").gsub(/[\s\t\r\n]/,'')
      return api_key
    end

    def initialize(app = nil, token = nil)
      super app
      @oauth2_token = token
    end

  end
end

Faraday::Request.register_middleware oauth2_refresh: FaradayMiddleware::OAuth2Refresh
```

**Comment by ntippie** [comment]
I drafted something up [here](https://github.com/minimul/qbo_api/compare/master...ntippie:oauth2refresh). I haven't tested it yet. This one shouldn't add any dependencies. It doesn't include any methods for the OAuth2 flow up until you get the refresh token. Interested in hearing if y'all think we should include them, and what you think about storing the expiry time versus refreshing on HTTP 401.

**Comment by bf4** [comment]
@ntippie I'll take a look.  I'd have some comments on code review.

I also have code for the authorization flow.

**Comment by minimul** [comment]
@ntippie Looks great. I would agree with @bf4 on triggering on a 401 because of Intuit's every changing refresh and token expiry durations.

@bf4 If you could put your code in a branch(es) so it can be diff'd and read easy. Also, I am going to hold to a Ruby 2.2.x minimum so no `dig()` and "safe nav." `&.` code.

**Comment by bf4** [comment]
@minimul yeah, just posting here as a kind of teaser... I have some refactoring of lib/qbo_api.rb etc that I'd like to do to make customization easier, as well as select different auth strategies

**Comment by JohnSmall** [comment]
As you're using the Rack::OAuth2 gem, at least in the example code. Then getting a new access token is easy. assuming you've set the rack client as in the example app with

     client = Rack::OAuth2::Client.new( etc...)
     
then you just do

     
       client.refresh_token = my_saved_refresh_token
       response = client.access_token!


You can then take apart the response to get the new access token and new refresh token.

I can see why people are suggesting acres of code using Faraday. The documentation for the rack-oauth2 gem is "read the source Luke" and it takes quite a lot of reading the source before you find the line that tells you how to refresh the token. But it is quite simple once you know.

**Comment by JohnSmall** [comment]
I came back to this after having switched to something else. I found that the rack-oauth2 gem, used in the example app does everything you need to refresh the token. See comment above. 

I'm now storing my tokens in an ActiveRecord object, along with the expiry time, so that when I request the access token it checks the expiry time and if it needs a new access token it goes and gets one and saves it and the new refresh token in the record. 

Storing tokens in AR does have a potential problem that it all happens in the context of a transaction. Which might get rolled back. In the past I've used yaml files or AWS DynamoDB tables to avoid transaction problems for tokens. But quite honestly it's more of a pain that it's worth, so for now I'm doing the easy thing and putting data in AR. 

I'll post my code if people think it's worth it.

**Comment by minimul** [comment]
Thanks for the that research.

Yes, I am interested to see your code.

**Comment by JohnSmall** [comment]
Ok, I'm in the process of implementing web hooks, as soon as I've done that I'll send my updates for refresh token. I've a question on web hooks, I'll put that in a new issue.

**Comment by JohnSmall** [comment]
I've been implementing the QBO stuff in a major project. Now I'm using it in earnest I really appreciate all the work you've done putting this together. It's a major task. Well done.

I was thinking of changing the QboApi methods so that you could pass in a method to be called whenever you get back a 401 and need to use the refresh token to get a new access token. 

I've not got around to that yet. At the moment I just have a simple class which stores the refresh and access tokens in AWS DynamoDB, so it's outside of any transactions, or testing rollbacks. Then there's a simple class level method #access_token that either pulls the access token from DynamoDB or if it has less than 30 seconds to run, it gets a new one, saves it and returns it. 

In short, you don't need to do anything inside the QboApi if you can store the tokens somewhere and check the time to expiry.

**Comment by minimul** [comment]
I'd like to see your code around this.

Also, I think it is fine to pass in a method to be called when getting back a 401. So if you want to put in a PR I will gladly receive it.

**Comment by Innysa** [comment]
Hi @minimul. How is progress on this issues? Is it in progress and will be delivered soon or it's better to write something myself? Is delivery postponed?

**Comment by bf4** [comment]
@Innysa It doesn't appear that there are any active PRs for put it in the gem, but there's plenty of examples in these issues.

**Comment by minimul** [comment]
I'd like to put in an official (but only if we have a consensus) solution soon.

/cc @ntippie

**Comment by ntippie** [comment]
From what I gather, the proposed solutions are:

- Using Faraday middleware (see my [untested branch](https://github.com/minimul/qbo_api/compare/master...ntippie:oauth2refresh))
- Using [bf4/safety_dance](https://github.com/bf4/safety_dance) as pattern or dependency
- Using [rack-oauth2](https://github.com/nov/rack-oauth2) for authentication

I like the idea of moving the authentication out of the gem, but we would need to also find a dependency for Oauth1, right? @minimul, what was your thinking about initially doing authentication directly in `qbo_api`?

**Comment by minimul** [comment]
@bf4 Are you using a similar renewal solution to what you posted last or did you leverage the [safety_dance](https://github.com/bf4/safety_dance) gem since?

**Comment by bf4** [comment]
Both actually

**Comment by minimul** [comment]
@JohnSmall Can you share what you are doing around renewal?

**Comment by minimul** [comment]
Intuit did release [this lib](https://github.com/intuit/oauth-rubyclient) somewhat recently.

**Comment by minimul** [comment]
I am going to go ahead with @ntippie 's Faraday middleware approach as the official library OAuth2 renewal solution.

@ntippie If you can post a PR and then I'll also be able to jump in since I must get a client switched over to OAuth2 by the end of month. Thanks.

**Comment by mikeyduece** [comment]
Hi! This is great and just what I was looking for :D I'm going to try to implement in my project now since it's not merged yet. That being said, I'm working on an app that I'm trying to incorporate with quickbooks. I'm using the `omniauth-quickbooks-oauth2`  gem for my authentication. I'm *not* using it to log into the application, only to associate with a user.

My question is, how are you handling an expired `access_token` apart from the `refresh_token`? I'm reading that the original access token that is returned after authenticating, will expire if not used within 100 days.

**Comment by minimul** [comment]
Hey @mikeyduece. 

Yes, we are going to have to implement a proc that handles a new access_token.

I'd recommend you wait until @ntippie and I discuss it more and update the branch.

With regards to OAuth2 libraries: API requests within the gem (`.get, .create, .update, etc`) will use [Faraday OAuth2 Middleware](https://github.com/lostisland/faraday_middleware/wiki/Oauth) but for Intuit login, Connect to QuickBooks functionality, revoke, just associating users, etc. `omniauth-quickbooks-oauth2` could be a good choice and I do use it on a client's QuickBooks Store app but only for Intuit login functionality. For other miscellaneous OAuth2 calls like the 'Connect to QuickBooks' button, revoke, etc. Rack::OAuth2 might be best because it has a closer-to-the-metal experience.

**Comment by mikeyduece** [comment]
I'm attempting to follow the example in your connection file. 
```
connection = build_connection('https://oauth.platform.intuit.com', headers: { 'Accept' => 'application/json' }) do |conn|
      conn.basic_auth(client_id, client_secret)
      conn.request :oauth2, 'token'
      conn.headers['Content-Type'] = 'application/x-www-form-urlencoded'
      conn.use FaradayMiddleware::ParseJson, parser_options: { symbolize_names: true }
      conn.use Faraday::Response::RaiseError
      conn.adapter Faraday.default_adapter
    end

    raw_response = connection.post { |req|
      req.body = { grant_type: :refresh_token, refresh_token: refresh_token }
      req.url '/oauth2/v1/tokens/bearer'
    }
```
But I keep getting `undefined method `request=' for #<Faraday::Connection:0x00007f95fbf6c8c0> Did you mean? request`

Sorry to bother about this. I'm just at a loss on how to best reup the refresh token

**Comment by ntippie** [comment]
Are you sure your code isn't running `conn.request = :oauth2, 'token'`? The method is `request`, not `request=`, as your error indicates.

Our planned OAuth2 support will involve you adding a `refresh_proc` and optionally an `access_proc` to each QboApi instance, which will be called when they change.

There is some work being done [here](https://github.com/minimul/qbo_api/compare/oauth2-refresh-mod), but it is unfinished and still being tested.

**Comment by mikeyduece** [comment]
I'm sure. I think maybe when I was testing from that PR, I didn't include enough. Thanks for the link, I will definitely give it another go on Monday.  Out of curiosity, when you're handling the refresh, how do you take into account that the refresh token, while good for '100days', expires after 30minutes? I've was thinking that I'm going to need to use the OmniAuth::Oauth2 library to essentially relog/reauth behind the scenes for each request if the `expires_at` is < Time.now

**Comment by ntippie** [comment]
I found [this video](https://www.youtube.com/watch?v=01dvFRGyA9Y) helpful. The refresh token lasts 101 days, but the time to expiration gets reset as you make requests. However, you may get a new refresh token when you request an access token.

The way we're doing this is that you pass a refresh_proc to `QboApi#new`, which will be called when there is a refresh token received. There is Faraday middleware that will automatically get a new access token and retry a request if needed.

**Comment by mikeyduece** [comment]
Thank you! I was able to hack together something from the links you sent until this is merged. But I was able to get the refresh token 😄 I'll put a todo in my project to refactor once this is merged. Thanks again for all of your help and *super* fast responses 💯

**Comment by mikeyduece** [comment]
Sorry to keep bothering you, you all have been so helpful. 
Today I got an invalid grant error from Faraday::ClientError. I wasn't able to refresh the token until after I reauthed with the omniauth gem. I haven't had any issues  since my last comment here 3wks ago with the refresh, until today. I fixed it, but was wondering if you had any insight on how to avoid that. My users authorize once in my app, and then the refresh is called as needed in the app under the hood. I'm thinking that a temp fix could be to set the status of the credential model to inactive and force the user,  through a pop up, to reconnect their quickbooks account, but I don't think that'd be great ux.

**Comment by bf4** [comment]
I run a job in the background to refresh every 10 minutes or so

**Comment by minimul** [comment]
@bf4 You are also catching 401s and renewing and replaying the request correct? 
Or is your app purely getting by with a background job?

**Comment by bf4** [comment]
@minimul We *do* catch 401s, refresh the token, and retry 
but that is kind of rare since we just refresh regularly. the token changes more regularly than the docs say.
I added the regular refresh awhile ago, but it seemed a better experience than having the possibility the token the expire if enough time passed between quickbooks api requests

**Comment by wheresrobert** [comment]
Hello!

We're getting ever closer to the deadline for upgrading to OAuth 2.0.

Are the code changes required to refresh tokens close to being production ready?

**Comment by wheresrobert** [comment]
> 
> 
> @wheresrobert I've been using OAuth 2.0 using this lib since the beginning. Worse case scenario, batteries aren't included and you do what everyone else is doing

I understand that I could do it myself, but I was curious to see if the work had already been done and is close to ready. If it's not, that's fine.

**Comment by minimul** [comment]
@wheresrobert Yes, we are in the process of a "batteries included"(@bf4) addition around token renewal. I am going to start investigating more next week. That said – no definite promises – there is a possibly that the token renewal addition will not be included if I think it is too complicated.

Stay tuned.

/cc @ntippie

**Comment by minimul** [comment]
I wanted to announce that I will not be considering a comprehensive "batteries included" library addition around OAuth2 token renewal.

I briefly tried the middleware solution (see above) without much success. After some digging around I thought the solution to be fairly complex and also started to groan at the idea of supporting it. Moreover, because of its complexities, I have a gut feeling including such a thing will degrade the overall gem quality and approachability.

I will have a series of screencasts soon on an approach to OAuth2 renewal and persistence. This approach is similar to @bf4 code samples above. It does involve wrapping all (or almost all) API requests in a Ruby exception hierarchy (which is the proper thing to do anyway) so it may require some significant refactoring.

The "middleware" approach appears to have great promise and its proper implementation means that you'll have to do little to no refactoring, therefore, I'll have no problem officially endorsing in the README a quality fork around this approach.

**Comment by minimul** [comment]
Note: I updated the README with an article where I discuss 2 possible renewal and persistence strategies.

Any feedback to those strategies as well as recommending alternative schemes are welcome on this thread.

**Comment by minimul** [comment]
Update: I have a question on a [interesting help.developer.intuit thread](https://help.developer.intuit.com/s/question/0D50f00006Kfd2XCAR/refreshtokenasync-returning-invalidgrant?t=1572537386934) in which I'm waiting for a response. 

Here is a [screenshot](https://drive.google.com/file/d/16kbZ4qeH6bY-YC5_0Ey6mM3davPckJY1/view?usp=sharing) of the thread as well because their forum app is not the best.

**Comment by minimul** [comment]
This is a reposting from the thread mentioned in the last post.
___
Update: I posted a longer treatise on this matter. 



https://minimul.com/quickbooks-online-api-oauth2-renewal-you-are-going-to-need-a-scheduled-background-job.html



Any corrections, rebukes or feedback is very appreciated.

**Comment by wheresrobert** [comment]
> This is a reposting from the thread mentioned in the last post.
> 
> Update: I posted a longer treatise on this matter.
> 
> https://minimul.com/quickbooks-online-api-oauth2-renewal-you-are-going-to-need-a-scheduled-background-job.html
> 
> Any corrections, rebukes or feedback is very appreciated.

I read the response from Intuit in a totally different way.

I read it as that the refresh token expires after 24 hours *if* you used it in a refresh call.

So it's 101 days *or* 24 hours if you refresh, whichever comes first. If you aren't refreshing your access token, then the refresh token is good for the 101 days.

This is why you need to save the refresh token you get along with the new access token at the time of refresh.

If I'm misunderstanding your assertion or you've seen something different happen, then I apologize up front.

**Comment by minimul** [comment]
@wheresrobert Thanks for responding.

Yes, that is what I understand as well.

My point is that you must implement – at minimum – a daily background job to guarantee that at least one access token API call is done within the 24 hour period to avoid an invalid_grant from its refresh token being past due (past due = no API call within the past 24 hours).

Yes, one is always persisting the refresh token after a successful access token renewal call.

Therefore, this "catch 401 error and retry" pattern (a pattern that Intuit has been championing as a "be all end all" renewal strategy) is not needed and instead a simple 10 or 15 minute scheduled background job is likely all you need. I say "likely" as I don't actually have any of my clients switched to OAuth2 in production as yet.

**Comment by wheresrobert** [comment]
> @wheresrobert Thanks for responding.
> 
> Yes, that is what I understand as well.
> 
> My point is that you must implement – at minimum – a daily background job to guarantee that at least one access token API call is done within the 24 hour period to avoid an invalid_grant from its refresh token being past due (past due = no API call within the past 24 hours).
> 
> Yes, one is always persisting the refresh token after a successful access token renewal call.
> 
> Therefore, this "catch 401 error and retry" pattern (a pattern that Intuit has been championing as a "be all end all" renewal strategy) is not needed and instead a simple 10 or 15 minute scheduled background job is likely all you need. I say "likely" as I don't actually have any of my clients switched to OAuth2 in production as yet.

I don't think that's accurate, but I will certainly verify that soon.

**Comment by itaisoudry** [comment]
I had to implement the refreshing process using a db that saves the refresh token.
Every time I want to make an api call ( well, at least every session ) I'm trying to refresh my access token, If it works and I get back a new refresh token, I update it in the db.

I don't think you need a background process, but just a safe place to store the token in the project \ lib.

**Comment by minimul** [comment]
@wheresrobert Appreciate it. I am very open to being corrected here.

**Comment by minimul** [comment]
@itaisoudry Are you always making at least one API call within a 24 hour period.

**Comment by minimul** [comment]
I just checked my app in development where I have OAuth2 implemented.

I made API calls last week and by my logic over the weekend with no API calls being made the refresh token should have expired, therefore, not allowing access token renewal – but it didn't do that and renewal worked so my understanding must be faulty.

**Comment by itaisoudry** [comment]
> @itaisoudry Are you always making at least one API call within a 24 hour period.

No. My code is part of a server that serves our registration system, I use the API only when I need to, usually more than 24 hours...

**Comment by minimul** [comment]
@itaisoudry Thanks.

I'll have to dig in more to see why I am getting regular invalid_grants.

**Comment by wheresrobert** [comment]
Is it possible that you're running into the other mentioned aspect in Intuit's reply?

e.g. Are you getting into a state where you call refresh token a bunch of times and maybe end up storing one of the earlier ones and not just the last one?

**Comment by minimul** [comment]
@wheresrobert `re: e.g.` => Not that scenario no but I'm going to dig in more this week.

**Comment by bf4** [comment]
We've had a background job running in prod regularly refreshing the token since before going live with our integration. (We've only had an OAuth 2 integration).

Intuit docs are misleading and this works and is fine.

I check the tokens after refresh and regularly notify myself

> "[Quickbooks] Changing refresh token even though expiry=8681415"
> "[Quickbooks] Changing access token even though expiry=3600 refresh_token_expires_in=8681415"

looks like it changes every 30 minutes or so

I sort of don't care. The background job is cheap and not need-ing to re-authenticate is the benefit.  Additional benefit is because the token is basically always fresh, normal requests are less likely to need to re-try due to the token being expired.

**Comment by nimisha84** [comment]
@wheresrobert is right. The refresh token expiry is 24hrs or 101days whichever comes first. Say, if I get  Acces Token 1(AT1) and Refresh Token 1(RT1) at 1pm on Monday. Then if my sync job runs and tests an api call(Say companyinfo) with AT1 on Thu, then I will get a 401. So, my job/service will do a refresh token call using RT1. RT1 will not expire on Thu as per above rule. You will get a new AT2 and RT2 on this call. Make sure to save both values everytime. As long as you use latest Refresh token, you will not get invalid grant errors.

**Comment by minimul** [comment]
Note: Just updated the README with "part 2" of my approach to implementing a "Step 1-5" renewal and persistence pattern.

**Comment by minimul** [comment]
Note: Just updated the README with "part 3" – Using the Migration API.

**Comment by minimul** [comment]
OAuth2 token renewal will not be baked into this gem. See README for solutions.
